File: drawing-java/pitchfork.js
class BasePitchfork {
    constructor(startX, startY, middleX, middleY, endX, endY) {
        this.startX = startX;
        this.startY = startY;
        this.middleX = middleX;
        this.middleY = middleY;
        this.endX = endX;
        this.endY = endY;
        this.lines = this.calculateLines();
    }

    moveStart(newX, newY) {
        const dx = newX - this.startX;
        const dy = newY - this.startY;
        this.startX = newX;
        this.startY = newY;
        this.moveAll(dx, dy);
    }

    moveMiddle(newX, newY) {
        const dx = newX - this.middleX;
        const dy = newY - this.middleY;
        this.middleX = newX;
        this.middleY = newY;
        this.moveAll(dx, dy);
    }

    moveEnd(newX, newY) {
        const dx = newX - this.endX;
        const dy = newY - this.endY;
        this.endX = newX;
        this.endY = newY;
        this.moveAll(dx, dy);
    }

    moveAll(dx, dy) {
        this.lines = this.lines.map(line => ({
            startPoint: {
                x: line.startPoint.x + dx,
                y: line.startPoint.y + dy
            },
            endPoint: {
                x: line.endPoint.x + dx,
                y: line.endPoint.y + dy
            }
        }));
    }

    adjustAngle(newEndX, newEndY) {
        const dx = this.endX - this.startX;
        const dy = this.endY - this.startY;
        const baseAngle = Math.atan2(dy, dx);
        
        const newDx = newEndX - this.startX;
        const newDy = newEndY - this.startY;
        const newAngle = Math.atan2(newDy, newDx);
        
        const rotationAngle = newAngle - baseAngle;
        
        [this.middleX, this.middleY] = this.rotatePoint(
            this.startX, this.startY,
            this.middleX, this.middleY,
            rotationAngle
        );
        
        [this.endX, this.endY] = [newEndX, newEndY];
        this.lines = this.calculateLines();
    }

    rotatePoint(centerX, centerY, x, y, angle) {
        const dx = x - centerX;
        const dy = y - centerY;
        const newX = centerX + dx * Math.cos(angle) - dy * Math.sin(angle);
        const newY = centerY + dx * Math.sin(angle) + dy * Math.cos(angle);
        return [newX, newY];
    }
}

class StandardPitchfork extends BasePitchfork {
    calculateLines() {
        const lines = [];
        
        // Handle line
        lines.push({
            startPoint: { x: this.startX, y: this.startY },
            endPoint: { x: this.middleX, y: this.middleY }
        });

        // Median line
        const medianX = (this.startX + this.endX) / 2;
        const medianY = (this.startY + this.endY) / 2;
        lines.push({
            startPoint: { x: this.middleX, y: this.middleY },
            endPoint: { x: medianX, y: medianY }
        });

        // Parallel lines
        const dx = this.endX - this.startX;
        const dy = this.endY - this.startY;
        lines.push({
            startPoint: { x: this.middleX + dx, y: this.middleY + dy },
            endPoint: { x: this.endX, y: this.endY }
        });

        return lines;
    }
}

class ModifiedSchiffPitchfork extends BasePitchfork {
    calculateLines() {
        const lines = [];
        
        // Modified handle
        const handleMidX = (this.startX + this.middleX) / 2;
        const handleMidY = (this.startY + this.middleY) / 2;
        lines.push({
            startPoint: { x: this.startX, y: this.startY },
            endPoint: { x: handleMidX, y: handleMidY }
        });

        // Modified median
        lines.push({
            startPoint: { x: handleMidX, y: handleMidY },
            endPoint: { x: this.endX, y: this.endY }
        });

        // Modified parallel lines
        const dx = this.endX - handleMidX;
        const dy = this.endY - handleMidY;
        lines.push({
            startPoint: { x: this.middleX + dx, y: this.middleY + dy },
            endPoint: { x: this.endX + dx, y: this.endY + dy }
        });

        return lines;
    }
}

class SchiffPitchfork extends BasePitchfork {
    calculateLines() {
        const lines = [];
        
        // Schiff handle
        lines.push({
            startPoint: { x: this.startX, y: this.startY },
            endPoint: { x: this.middleX, y: this.middleY }
        });

        // Schiff median
        const medianX = (this.startX + this.middleX) / 2;
        const medianY = (this.startY + this.middleY) / 2;
        lines.push({
            startPoint: { x: medianX, y: medianY },
            endPoint: { x: this.endX, y: this.endY }
        });

        // Schiff parallel lines
        const dx = this.endX - this.startX;
        const dy = this.endY - this.startY;
        lines.push({
            startPoint: { x: this.middleX + dx, y: this.middleY + dy },
            endPoint: { x: this.endX, y: this.endY }
        });

        return lines;
    }
}

class InsidePitchfork extends BasePitchfork {
    calculateLines() {
        const lines = [];
        
        // Inside handle
        lines.push({
            startPoint: { x: this.startX, y: this.startY },
            endPoint: { x: this.middleX, y: this.middleY }
        });

        // Inside median
        const medianX = (this.startX + this.endX) / 2;
        const medianY = (this.startY + this.endY) / 2;
        lines.push({
            startPoint: { x: this.middleX, y: this.middleY },
            endPoint: { x: medianX, y: medianY }
        });

        // Inside parallel lines (internal tines)
        const dx = (this.endX - this.startX) * 0.5;
        const dy = (this.endY - this.startY) * 0.5;
        lines.push({
            startPoint: { x: this.middleX - dx, y: this.middleY - dy },
            endPoint: { x: this.middleX + dx, y: this.middleY + dy }
        });

        return lines;
    }
}

class BarleyPitchfork extends BasePitchfork {
    calculateLines() {
        const lines = [];
        
        // Barley handle
        lines.push({
            startPoint: { x: this.startX, y: this.startY },
            endPoint: { x: this.middleX, y: this.middleY }
        });

        // Barley median with warning line
        const medianX = (this.startX + this.endX) / 2;
        const medianY = (this.startY + this.endY) / 2;
        lines.push({
            startPoint: { x: this.middleX, y: this.middleY },
            endPoint: { x: medianX, y: medianY }
        });

        // Barley parallel lines with warning lines
        const dx = this.endX - this.startX;
        const dy = this.endY - this.startY;
        
        // Main parallel lines
        lines.push({
            startPoint: { x: this.middleX + dx, y: this.middleY + dy },
            endPoint: { x: this.endX, y: this.endY }
        });

        // Warning lines (0.618 and 1.618 extensions)
        const phi = 1.618;
        lines.push({
            startPoint: { x: this.middleX + dx * phi, y: this.middleY + dy * phi },
            endPoint: { x: this.endX + dx * (phi - 1), y: this.endY + dy * (phi - 1) }
        });
        
        lines.push({
            startPoint: { x: this.middleX + dx * 0.618, y: this.middleY + dy * 0.618 },
            endPoint: { x: this.endX - dx * 0.382, y: this.endY - dy * 0.382 }
        });

        return lines;
    }
}

export { 
    BasePitchfork,
    StandardPitchfork,
    ModifiedSchiffPitchfork,
    SchiffPitchfork,
    InsidePitchfork,
    BarleyPitchfork
};

File: drawing-logic/pitchfork.py
import math

class Pitchfork:
    def __init__(self, start_x, start_y, middle_x, middle_y, end_x, end_y):
        self.start_x = start_x
        self.start_y = start_y
        self.middle_x = middle_x
        self.middle_y = middle_y
        self.end_x = end_x
        self.end_y = end_y
        self.lines = self._calculate_lines()

    def _calculate_lines(self):
        # Calculate the lines of the pitchfork
        start_x, start_y = self.start_x, self.start_y
        middle_x, middle_y = self.middle_x, self.middle_y
        end_x, end_y = self.end_x, self.end_y

        lines = []
        lines.append(TrendLine(start_x, start_y, middle_x, middle_y))
        
        # Calculate the median line
        median_x = (start_x + end_x) / 2
        median_y = (start_y + end_y) / 2
        lines.append(TrendLine(middle_x, middle_y, median_x, median_y))

        # Calculate the top and bottom lines
        dx = end_x - start_x
        dy = end_y - start_y
        
        lines.append(TrendLine(middle_x + dx, middle_y + dy, end_x, end_y))
        
        return lines

    def draw(self):
        # Draw all the lines of the pitchfork
        path = ""
        for line in self.lines:
            path += f"M{line.start_x},{line.start_y} L{line.end_x},{line.end_y} "
        return path

    def move_start(self, new_x, new_y):
        # Move the start point of the pitchfork
        dx = new_x - self.start_x
        dy = new_y - self.start_y
        self.start_x = new_x
        self.start_y = new_y
        self.move_all(dx, dy)

    def move_middle(self, new_x, new_y):
        # Move the middle point of the pitchfork
        dx = new_x - self.middle_x
        dy = new_y - self.middle_y
        self.middle_x = new_x
        self.middle_y = new_y
        self.move_all(dx, dy)

    def move_end(self, new_x, new_y):
        # Move the end point of the pitchfork
        dx = new_x - self.end_x
        dy = new_y - self.end_y
        self.end_x = new_x
        self.end_y = new_y
        self.move_all(dx, dy)

    def move_all(self, dx, dy):
        # Move all the lines of the pitchfork
        for line in self.lines:
            line.start_x += dx
            line.start_y += dy
            line.end_x += dx
            line.end_y += dy

    def adjust_angle(self, new_end_x, new_end_y):
        # Adjust the angle of the entire pitchfork
        
        # Calculate the angle of the base line
        dx = self.end_x - self.start_x
        dy = self.end_y - self.start_y
        base_angle = math.atan2(dy, dx)
        
        # Calculate the length of the base line
        base_length = math.sqrt(dx**2 + dy**2)
        
        # Calculate the new angle
        new_dx = new_end_x - self.start_x
        new_dy = new_end_y - self.start_y
        new_angle = math.atan2(new_dy, new_dx)
        
        # Calculate the rotation angle
        rotation_angle = new_angle - base_angle
        
        # Rotate the middle and end points
        self.middle_x, self.middle_y = self._rotate_point(self.start_x, self.start_y, self.middle_x, self.middle_y, rotation_angle)
        self.end_x, self.end_y = self._rotate_point(self.start_x, self.start_y, new_end_x, new_end_y, rotation_angle)
        
        # Recalculate the lines
        self.lines = self._calculate_lines()

    def _rotate_point(self, center_x, center_y, x, y, angle):
        # Rotate a point around a center
        dx = x - center_x
        dy = y - center_y
        new_x = center_x + dx * math.cos(angle) - dy * math.sin(angle)
        new_y = center_y + dx * math.sin(angle) + dy * math.cos(angle)
        return new_x, new_y

class NormalPitchfork(Pitchfork):
    def __init__(self, start_x, start_y, middle_x, middle_y, end_x, end_y):
        super().__init__(start_x, start_y, middle_x, middle_y, end_x, end_y)

class SchiffPitchfork(Pitchfork):
    def __init__(self, start_x, start_y, middle_x, middle_y, end_x, end_y):
        super().__init__(start_x, start_y, middle_x, middle_y, end_x, end_y)

    def _calculate_lines(self):
        # Calculate the lines of the schiff pitchfork
        start_x, start_y = self.start_x, self.start_y
        middle_x, middle_y = self.middle_x, self.middle_y
        end_x, end_y = self.end_x, self.end_y

        lines = []
        lines.append(TrendLine(start_x, start_y, middle_x, middle_y))
        
        # Calculate the median line
        median_x = (start_x + middle_x) / 2
        median_y = (start_y + middle_y) / 2
        lines.append(TrendLine(median_x, median_y, end_x, end_y))

        # Calculate the top and bottom lines
        dx = end_x - start_x
        dy = end_y - start_y
        
        lines.append(TrendLine(middle_x + dx, middle_y + dy, end_x, end_y))
        
        return lines

class ModifiedPitchfork(Pitchfork):
    def __init__(self, start_x, start_y, middle_x, middle_y, end_x, end_y):
        super().__init__(start_x, start_y, middle_x, middle_y, end_x, end_y)

    def _calculate_lines(self):
        # Calculate the lines of the modified pitchfork
        start_x, start_y = self.start_x, self.start_y
        middle_x, middle_y = self.middle_x, self.middle_y
        end_x, end_y = self.end_x, self.end_y

        lines = []
        lines.append(TrendLine(start_x, start_y, middle_x, middle_y))
        
        # Calculate the median line
        median_x = (middle_x + end_x) / 2
        median_y = (middle_y + end_y) / 2
        lines.append(TrendLine(middle_x, middle_y, median_x, median_y))

        # Calculate the top and bottom lines
        dx = end_x - start_x
        dy = end_y - start_y
        
        lines.append(TrendLine(middle_x + dx, middle_y + dy, end_x, end_y))
        
        return lines

class InsidePitchfork(Pitchfork):
    def __init__(self, start_x, start_y, middle_x, middle_y, end_x, end_y):
        super().__init__(start_x, start_y, middle_x, middle_y, end_x, end_y)

    def _calculate_lines(self):
        # Calculate the lines of the inside pitchfork
        start_x, start_y = self.start_x, self.start_y
        middle_x, middle_y = self.middle_x, self.middle_y
        end_x, end_y = self.end_x, self.end_y

        lines = []
        lines.append(TrendLine(start_x, start_y, middle_x, middle_y))
        
        # Calculate the median line
        median_x = (start_x + end_x) / 2
        median_y = (start_y + end_y) / 2
        lines.append(TrendLine(middle_x, middle_y, median_x, median_y))

        # Calculate the top and bottom lines
        dx = end_x - start_x
        dy = end_y - start_y
        
        lines.append(TrendLine(middle_x - dx, middle_y - dy, end_x, end_y))
        
        return lines

